File: 394-decode-string/decode-string.java
class Solution {
    int index = 0;
    public String decodeString(String s) {
        StringBuilder res = new StringBuilder();
        while (index < s.length() && s.charAt(index) != ']') {
            char ch = s.charAt(index);
            if(!Character.isDigit(ch)) {
                res.append(ch);
                index++;
            } else {
                int k = 0;
                while (index < s.length() && Character.isDigit(s.charAt(index))) {
                    k = k * 10 + (s.charAt(index) - '0');
                    index++;
                }
                index++;
                String decodeString = decodeString(s);
                index++;
                while (k > 0) {
                    res.append(decodeString);
                    k--;
                }
            }
        }
        return res.toString();
    }
}

File: 394-decode-string/decode-string.py
class Solution(object):
    index = 0
    def decodeString(self, s):
        result = ""
        while self.index < len(s) and s[self.index] != ']':
            if not s[self.index].isdigit():
                result += s[self.index]
                self.index += 1
            else:
                k = 0
                while self.index < len(s) and s[self.index].isdigit():
                    k = (k * 10) + (int(s[self.index]))
                    self.index += 1
                self.index += 1
                rem = self.decodeString(s)
                self.index += 1
                while k > 0:
                    result = result + rem
                    k -= 1
        return result